fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/digitalcash/HistoryListAdapter.java

```diff
@@ -0,0 +1,242 @@
+import androidx.recyclerview.widget.RecyclerView;
+
+import com.github.dedis.popstellar.R;
+import com.github.dedis.popstellar.model.objects.InputObject;
+import com.github.dedis.popstellar.model.objects.digitalcash.TransactionObject;
+import com.github.dedis.popstellar.model.objects.security.PoPToken;
+import com.github.dedis.popstellar.model.objects.security.PublicKey;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class HistoryListAdapter extends RecyclerView.Adapter<HistoryListAdapter.HistoryViewHolder> {
+  private static final String TAG = HistoryListAdapter.class.getSimpleName();
+  private List<TransactionHistoryElement> transactions;
+  private final Map<String, Boolean> expandMap;
+  Set<PublicKey> ownPublicKeys;
+  Set<String> ownPublicKeysHash;
+
+  public HistoryListAdapter(List<TransactionObject> transactionObjects, Set<PoPToken> ownTokens) {
+    if (transactionObjects == null) {
+      transactionObjects = new ArrayList<>();
+    }
+    this.ownPublicKeys = getPksFromTokens(ownTokens);
+    this.ownPublicKeysHash =
+        ownPublicKeys.parallelStream().map(PublicKey::computeHash).collect(Collectors.toSet());
+
+    this.transactions = buildTransactionList(transactionObjects);
+    expandMap = new HashMap<>();
+    transactions.forEach(
+        transactionHistoryElement -> expandMap.put(transactionHistoryElement.id, false));
+  }
+
+  @NonNull
+  @Override
+  public HistoryViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
+    View view =
+        LayoutInflater.from(parent.getContext())
+            .inflate(R.layout.history_transaction_layout, parent, false);
+    return new HistoryViewHolder(view);
+  }
+
+  @Override
+  public void onBindViewHolder(@NonNull HistoryViewHolder holder, int position) {
+    TransactionHistoryElement element = transactions.get(position);
+    if (element == null) {
+      Log.d(TAG, "element is null");
+      return;
+    }
+    String transactionId = element.getId();
+    Boolean expandObject = expandMap.get(transactionId);
+    boolean expand = expandObject != null && expandObject;
+
+    holder.detailLayout.setVisibility(expand ? View.VISIBLE : View.GONE);
+    holder.expandIcon.setRotation(expand ? 180f : 0f);
+    holder.transactionTypeTitle.setText(
+        element.isSender() ? R.string.digital_cash_sent : R.string.digital_cash_received);
+
+    holder.transactionTypeValue.setText(element.getValue());
+
+    holder.transactionIdValue.setText(transactionId);
+
+    holder.transactionProvenanceTitle.setText(
+        element.isSender() ? R.string.digital_cash_to : R.string.digital_cash_from);
+    holder.transactionProvenanceValue.setText(element.getSenderOrReceiver());
+
+    View.OnClickListener listener =
+        v -> {
+          Log.d(TAG, "transaction is " + transactionId + " position is " + position);
```

Comment:
Originally a debug line but does it make sense to leave it ?

Reply:
I would tend to say yes, since it can be useful to know when a transaction happens what it it's id